hardware/function_generators.py

```diff
@@ -159,56 +263,72 @@ class SRS_DS345():
     Attributes:
         frequency (float): The frequency in Hz. Also aliases to ``freq``
         voltage (float): The peak-to-peak voltage in volts. Also aliases to
-        ``volt`` waveform (str):  A string in the array
+        ``volt``
+        waveform (str):  A string in the array
             ``['SIN', 'SQU', 'RAMP', 'PULS', 'NOIS', 'DC', 'USER' ]``
+
         phase (float): The relative phase to other instruments synced through
         the 10 MHz reference clock.
         duty_cycle (float): The duty cycle of the square wave form in percent.
     """
     def __init__(self, visa_search_term):
-        rm = visa.ResourceManager()
-        self.inst = rm.open_resource(visa_search_term)
+         super(SRS_DS345, self).__init__(visa_search_term)
 
-    def identify(self):
-        """
-        Returns:
-            str: the response from the ``*IDN?`` GPIB query.
-        """
-        return self.inst.query('*IDN?')[:-1]
 
     @property
     def frequency(self):
         f = self.inst.query('FREQ?')
-
-        return float(f)
+        return float(f) * u.hertz
 
     @frequency.setter
     def frequency(self, val):
+        if(self.waveform_name == "NOIS"):
+            raise ValueError("Frequency must remain at 10 MHz when waveform is 'NOISE'")
+        elif(val<1 * u.microhertz):
+            raise ValueError("Minimum frequency is 1 µHz")
+        elif((self.waveform_name == "SIN" or self.waveform_name == "SQ") and
+              val>30.2 * u.megahertz):
+            raise ValueError("Maximum frequency is 30.2 MHz for %s waves" % waveform_name)
+        elif(self.waveform_name == "RAMP" and val>100 * u.kilohertz):
+            raise ValueError("Maximum frequency is 100 KHz for ramp waves")
         self.inst.write('FREQ %i' % val)
 
     # alias
     freq = frequency
 
     @property
     def voltage(self):
-        return float(self.inst.query('AMPL?')[:-3])
+        return float(self.inst.query('AMPL?')[:-3]) * u.volt
 
     @voltage.setter
     def voltage(self, val):
+# list of the max and mins of all the voltages you can have based on the waves

```

Comment:
Maybe these can go into some class/dictionary at some point. Put it in the issue tracker